File: examples/producer/src/handler.py
import json
import os
import random
import time
import uuid
from datetime import datetime, timezone

import boto3
from aws_lambda_powertools import Logger, Metrics, Tracer
from aws_lambda_powertools.metrics import MetricUnit
from aws_lambda_powertools.metrics import single_metric

logger = Logger()
tracer = Tracer()
metrics = Metrics()

QUEUE_URL = os.environ["QUEUE_URL"]
sqs = boto3.client("sqs")  # Tracer auto-patches boto3

HIGH_AMOUNT_CENTS = 100_000
POISON_RATE = 0.05
DEBUG_LOG_RATE = 0.1


@tracer.capture_method
def _classify_amount(amount: int) -> str:
    if amount < 0:
        return "poison"
    if amount >= HIGH_AMOUNT_CENTS:
        return "high"
    return "normal"


@tracer.capture_method
def _build_order() -> dict:
    if random.random() < POISON_RATE:
        amount = -random.randint(1, 100)
    else:
        amount = random.randint(100, 250_000)
    order = {
        "orderId": str(uuid.uuid4()),
        "customerId": str(uuid.uuid4()),
        "amountCents": amount,
        "createdAt": datetime.now(timezone.utc).isoformat(),
    }
    if random.random() < DEBUG_LOG_RATE:
        logger.debug("order_serialized", extra={"orderId": order["orderId"]})
    return order


@tracer.capture_method
def _send_order(order: dict) -> None:
    sqs.send_message(QueueUrl=QUEUE_URL, MessageBody=json.dumps(order))


@logger.inject_lambda_context(log_event=False)
@tracer.capture_lambda_handler
@metrics.log_metrics(capture_cold_start_metric=True)
def handler(event, _context):
    metrics.add_dimension(name="environment", value=os.environ.get("STAGE", "dev"))

    order = _build_order()
    shape = _classify_amount(order["amountCents"])

    tracer.put_annotation(key="orderId", value=order["orderId"])
    tracer.put_annotation(key="orderShape", value=shape)
    tracer.put_metadata(key="order", value=order)

    started = time.perf_counter()
    try:
        _send_order(order)
    except Exception:
        logger.exception("send_failed", extra={"orderId": order["orderId"]})
        metrics.add_metric(name="SendFailures", unit=MetricUnit.Count, value=1)
        raise
    elapsed_ms = (time.perf_counter() - started) * 1000

    payload_bytes = len(json.dumps(order).encode("utf-8"))
    metrics.add_metric(name="OrdersEmitted", unit=MetricUnit.Count, value=1)
    metrics.add_metric(name="PayloadBytes", unit=MetricUnit.Bytes, value=payload_bytes)
    metrics.add_metric(
        name="EmitLatencyMs", unit=MetricUnit.Milliseconds, value=elapsed_ms
    )

    with single_metric(
        name="OrdersByShape",
        unit=MetricUnit.Count,
        value=1,
        namespace=os.environ.get("POWERTOOLS_METRICS_NAMESPACE", ""),
    ) as m:
        m.add_dimension(name="orderShape", value=shape)

    if shape == "poison":
        logger.critical(
            "poison_emitted",
            extra={"orderId": order["orderId"], "amountCents": order["amountCents"]},
        )
    elif shape == "high":
        logger.warning(
            "high_amount",
            extra={
                "orderId": order["orderId"],
                "amountCents": order["amountCents"],
            },
        )

    logger.info(
        "order_emitted",
        extra={
            "orderId": order["orderId"],
            "orderShape": shape,
            "elapsedMs": round(elapsed_ms, 2),
        },
    )
    return {"orderId": order["orderId"]}

File: examples/producer/src/handler.ts
import { Logger as PowertoolsLogger } from "@aws-lambda-powertools/logger";
import {
  Metrics as PowertoolsMetricsCtor,
  MetricUnit,
} from "@aws-lambda-powertools/metrics";
import { Tracer as PowertoolsTracer } from "@aws-lambda-powertools/tracer";
import { SendMessageCommand, SQSClient } from "@aws-sdk/client-sqs";
import * as Effect from "effect/Effect";
import * as Metric from "effect/Metric";
import * as Schema from "effect/Schema";

import {
  counter,
  createHandler,
  histogram,
  PowertoolsLayer,
} from "effect-powertools";

const ptLogger = new PowertoolsLogger();
const ptTracer = new PowertoolsTracer();
const ptMetrics = new PowertoolsMetricsCtor();

const sqs = ptTracer.captureAWSv3Client(new SQSClient({}));

const QUEUE_URL = process.env.QUEUE_URL;
if (!QUEUE_URL) throw new Error("QUEUE_URL env var is required");

const HIGH_AMOUNT_CENTS = 100_000;
const POISON_RATE = 0.05;
const DEBUG_LOG_RATE = 0.1;

const TriggerEvent = Schema.Struct({
  source: Schema.optional(Schema.String),
});

interface Order {
  readonly orderId: string;
  readonly customerId: string;
  readonly amountCents: number;
  readonly createdAt: string;
}

class SendOrderError {
  readonly _tag = "SendOrderError";
  constructor(
    readonly props: { readonly orderId: string; readonly cause: string },
  ) {}
  toString() {
    return `SendOrderError(${this.props.orderId}): ${this.props.cause}`;
  }
}

const ordersEmitted = counter("OrdersEmitted", { unit: MetricUnit.Count });
const sendFailures = counter("SendFailures", { unit: MetricUnit.Count });
const ordersByShape = counter("OrdersByShape", { unit: MetricUnit.Count });
const payloadBytes = histogram(
  "PayloadBytes",
  [128, 256, 512, 1024, 2048],
  { unit: MetricUnit.Bytes },
);
const emitLatency = Metric.timer("EmitLatencyMs");

const classifyAmount = (amount: number): string => {
  if (amount < 0) return "poison";
  if (amount >= HIGH_AMOUNT_CENTS) return "high";
  return "normal";
};

const buildOrder = Effect.gen(function* () {
  const amount =
    Math.random() < POISON_RATE
      ? -(Math.floor(Math.random() * 100) + 1)
      : Math.floor(Math.random() * 250_000) + 100;
  const order: Order = {
    orderId: globalThis.crypto.randomUUID(),
    customerId: globalThis.crypto.randomUUID(),
    amountCents: amount,
    createdAt: new Date().toISOString(),
  };
  if (Math.random() < DEBUG_LOG_RATE) {
    yield* Effect.logDebug("order_serialized").pipe(
      Effect.annotateLogs({ orderId: order.orderId }),
    );
  }
  return order;
}).pipe(Effect.withSpan("buildOrder"));

const sendOrder = (order: Order, body: string) =>
  Effect.tryPromise({
    try: () =>
      sqs.send(
        new SendMessageCommand({ QueueUrl: QUEUE_URL, MessageBody: body }),
      ),
    catch: (error) =>
      new SendOrderError({ orderId: order.orderId, cause: String(error) }),
  }).pipe(
    Effect.withSpan("sqs.sendMessage", {
      attributes: { queueUrl: QUEUE_URL, orderId: order.orderId },
    }),
    Metric.trackDuration(emitLatency),
  );

const program = Effect.gen(function* () {
  const order = yield* buildOrder;
  const body = JSON.stringify(order);
  const shape = classifyAmount(order.amountCents);

  yield* Effect.annotateCurrentSpan("orderId", order.orderId);
  yield* Effect.annotateCurrentSpan("orderShape", shape);

  yield* sendOrder(order, body).pipe(
    Effect.tapError((error) =>
      Effect.zipRight(
        Metric.update(sendFailures, 1),
        Effect.logError("send_failed").pipe(
          Effect.annotateLogs({
            orderId: order.orderId,
            error: String(error),
          }),
        ),
      ),
    ),
  );

  yield* Metric.update(ordersEmitted, 1);
  yield* Metric.update(
    Metric.tagged(payloadBytes, "orderShape", shape),
    Buffer.byteLength(body, "utf8"),
  );
  yield* Metric.update(
    Metric.tagged(ordersByShape, "orderShape", shape),
    1,
  );

  if (shape === "poison") {
    yield* Effect.logFatal("poison_emitted").pipe(
      Effect.annotateLogs({
        orderId: order.orderId,
        amountCents: order.amountCents,
      }),
    );
  } else if (shape === "high") {
    yield* Effect.logWarning("high_amount").pipe(
      Effect.annotateLogs({
        orderId: order.orderId,
        amountCents: order.amountCents,
      }),
    );
  }

  yield* Effect.logInfo("order_emitted").pipe(
    Effect.annotateLogs({ orderId: order.orderId, orderShape: shape }),
  );

  return { orderId: order.orderId };
});

export const handler = createHandler(
  {
    schema: TriggerEvent,
    layer: PowertoolsLayer({
      logger: ptLogger,
      tracer: ptTracer,
      metrics: ptMetrics,
    }),
    serviceName: "producer",
  },
  (_event, _context) => program,
);
